twap/price/price.go

```diff
@@ -0,0 +1,48 @@
+package price
+
+import (
+	"fmt"
+	"sort"
+	"time"
+)
+
+type Price struct {
+	Value     float64   `json:"price"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func TWAP(samples []Price) (float64, error) {
+	if len(samples) == 0 {
+		return 0, fmt.Errorf("no samples provided")
+	}
+	if len(samples) == 1 {
+		return samples[0].Value, nil
+	}
+
+	// Sort samples from latest to earliest
+	sort.Slice(
+		samples, func(i, j int) bool {
+			return samples[i].Timestamp.After(samples[j].Timestamp)
+		},
+	)
+
+	var weightedSum, totalWeight float64
+
+	// IMPORTANT: The value of the last sample is not included in the calculation
```

Comment:
It shouldn't matter much since we don't know the true function that we are estimating but but since we are omitting the value of the last sample, we may want to use the midpoint of the range for the estimator.  That is, instead of using `prev.Value` on line 38, use `(prev.Value + next.Value) / 2`.